### DISCUSSION TITLE
V7 to V8 Upgrade Issue

### DISCUSSION THREAD

**Discussion by ankitjainm** [webdriverio/webdriverio#10616] [Technical Discussion]
I upgraded from Webdriver V7 to V8, while trying to run the tests, I am getting the below error, any idea on how to resolve it?

I tried with different "@wdio/mocha-framework" versions but it did not help

file:///Users/<User Name>/Documents/GitHub/<Project Name>/node_modules/@wdio/mocha-framework/build/common.js:116
[0-0]         throw new Error(`Module ${name} can't get loaded. Are you sure you have installed it?

**Comment by christian-bromann** [comment]
Can you provide the full terminal logs?

**Comment by ankitjainm** [comment]
Thanks for responding, sharing the complete log

file:///Users/<User Name>/Documents/GitHub/<Repository>/node_modules/@wdio/mocha-framework/build/common.js:116
0-0]         throw new Error(`Module ${name} can't get loaded. Are you sure you have installed it?\n` +
[0-0]               ^
[0-0] Error: Module tsconfig-paths/register can't get loaded. Are you sure you have installed it?
[0-0] Note: if you've installed WebdriverIO globally you need to install these external modules globally too!
[0-0]     at loadModule (file:///Users/<User Name>/Documents/GitHub/<Repository>/node_modules/@wdio/mocha-framework/build/common.js:116:15)

**Comment by christian-bromann** [comment]
> Error: Module tsconfig-paths/register can't get loaded.

v8 doesn't support this feature anymore. It doesn't seem you've completely updated all WebdriverIO packages correctly.

**Reply by ankitjainm** [reply]
May I know which feature is not supported, below is my package.json, May I know which package needs update or is missing

"devDependencies": {
    "@types/chai": "^4.3.5",
    "@types/mocha": "^10.0.1",
    "@types/node": "^20.3.1",
    "@typescript-eslint/eslint-plugin": "^5.60.0",
    "@typescript-eslint/parser": "^5.60.0",
    "@wdio/allure-reporter": "^8.11.0",
    "chromedriver": "^114.0.2",
    "edgedriver": "5.2.1",
    "eslint": "^8.43.0",
    "geckodriver": "^4.0.4",
    "husky": "^8.0.3",
    "lint-staged": "^13.2.2",
    "ms-chromium-edge-driver": "^0.5.1",
    "safaridriver": "0.0.5",
    "ts-node": "^10.9.1",
    "tsconfig-paths": "^4.2.0",
    "typescript": "^5.1.3",
    "wdio-chromedriver-service": "^8.1.1",
    "wdio-edgedriver-service": "^3.0.2",
    "wdio-intercept-service": "^4.4.0",
    "wdio-safaridriver-service": "^2.1.0"
  },
  "dependencies": {
    "@wdio/cli": "^8.0.9",
    "@wdio/junit-reporter": "8.4.0",
    "@wdio/local-runner": "^8.5.9",
    "@wdio/mocha-framework": "^8.5.9",
    "@wdio/sauce-service": "^8.11.2",
    "@wdio/selenium-standalone-service": "^8.11.0",
    "@wdio/spec-reporter": "^8.4.0",
    "axios": "^1.4.0",
    "chai": "^4.3.7",
    "dotenv": "^16.3.1",
    "mocha": "^10.2.0",
    "mocha-it-each": "^1.0.2",
    "node-fetch": "^3.3.1",
    "speakeasy": "^2.0.0",
    "tslint": "^6.1.2"
  },

**Reply by christian-bromann** [reply]
The package versions seems updated but were they installed? I think the easiest is if you could provide a reproducible example.

**Reply by ankitjainm** [reply]
Yes the packages got installed, I am also getting the below error, any idea about it

"Error: Module tsconfig-paths/register can't get loaded. Are you sure you have installed it?"

Also sharing the tsconfig file below

{
  "compilerOptions": {
    "target": "ES2019",
    "sourceMap": true,
    "outDir": "dist",
    "module": "commonjs",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "esModuleInterop": true,
    "baseUrl": ".",
    "typeRoots": [
      "./types",
      "node_modules/@types"
    ],
    "types": ["node", "webdriverio/async", "@wdio/mocha-framework", "chai", "wdio-intercept-service"],
    "noEmitOnError": true
  },
  "include": ["./**/*.ts"],
  "exclude": ["node_modules", "dist", "config"],
  "ts-node": {
    "files": true,
    "require": ["tsconfig-paths/register"]
  }
}

**Reply by christian-bromann** [reply]
@ankitjainm I can't say what is going on without a reproducible example.

**Reply by ankitjainm** [reply]
Can you please provide more context about reproducible example, happy to share more details on it.

**Reply by christian-bromann** [reply]
Please create a project where this problem can be reproduced by me via cloning it and calling a certain command. See also https://stackoverflow.com/help/minimal-reproducible-example

**Reply by ankitjainm** [reply]
Sure, I have shared the package.json, wdio.conf.ts and base.conf.ts file with you, I was having difficulty creating a complete project, hopefully it will work.

If possible, can you please check and let me know if the config is correct. I think the main issue is with compiler option mentioned in base.conf.ts, is this the correct way to point the compiler in Webdriver.io V8?

mochaOpts: {
        ui: "bdd",
        timeout: 1000000,
        //require: 'ts-node/register',
        compilers: ["tsconfig-paths/register"],
        retries: 1
    }

[Test1.zip](https://github.com/webdriverio/webdriverio/files/11859024/Test1.zip)

**Reply by ankitjainm** [reply]
Hello @christian-bromann I made some progress, created a new wdio project and updated the package.json and tsconfig.json with config, however after executing the test scripts, I am getting error 'Cannot find module' for all the files which are importing from other files. I also updated "type": "module" in package.json and reinstalled the node modules but still faced the same issue, any idea on how to resolve it?

**Comment by chamanpreet010** [comment]
Hi @ankitjainm I am facing same issue while upgrading webdriverIO from 7 to 8. Did you get any solution to it?

**Reply by ankitjainm** [reply]
sry did not

**Comment by andy250** [comment]
I faced similar issue (porting v7 to v8 using typescript). In the end I took this approach:
* install fresh wdio v8 project `npm init wdio@latest .` - follow all the wizard questions carefully
* make sure to include some "sample tests" (the init flow will ask you if you want them)
* check out new project's config files, sample pageobjects and sample test files
* apply the same configs to your existing project
* remember that v8:
  *  needs more imports e.g. `import { $, browser, expect } from '@wdio/globals'`
  * .ts files are imported using .js extension
```
import LoginPage from '../pageobjects/login.page.js'
import SecurePage from '../pageobjects/secure.page.js'
```
* important part in `tsconfig.json`
```
{
    "compilerOptions": {
        "types": [
            "node",
            "@wdio/globals/types",
            "expect-webdriverio",
            "@wdio/mocha-framework"
        ]
    }
}
```
Also note that v8 release notes state this:
> Lastly, we had to remove support for [tsconfig-paths](https://www.npmjs.com/package/tsconfig-paths) as we haven't found a way to get it working within an ESM context. We believe this integration hasn't been used much anyway and a lot of it is nowadays natively supported in TypeScript. Let us know if this assumption is wrong and you would like to see it being supported again.

So I am guessing this part of mocha config will not work anymore `compilers: ["tsconfig-paths/register"]`

**Reply by TheRealCleany** [reply]
Yeah, but I have and I know a lot of people do.
I've tried skipping 8 and go to 9, but that was even more of a disaster.
My async methods even stopped working. I mean, come on!

**Reply by christian-bromann** [reply]
Hey @TheRealCleany , thanks for your comment. Just to give you the perspective of a maintainer: a project like WebdriverIO has tons of little features and nuances that makes it impossible to track especially when it comes to interoperability to 3rd party dependencies. The maintainer team would love to offer a better experience upgrading from one version to the other but this often fails due to lack of time and resources. Note that this is an open source project maintained by volunteers. You as the user are as much responsible to improve the project like any of the contributors are.

With that in mind, let's identify your issues, ensure that we have documented certain changes properly so others can have a better experience 😉
